### PR TITLE
Don't cancel aborted jobs immediately

### LINKED ISSUE #11084: Occasional `IOException`/`InterruptedException` during startup
Screenshot from 2024-09-16 11-58-16
Screenshot from 2024-09-16 11-59-31

Our execution can be interrupted but it should only happen when there is another one in the queue scheduled.

### PR DESCRIPTION
**Pull request by hubertp**
### Pull Request Description

Rather than cancelling Futures that capture jobs' logic, this change introduces a two-level system:
- interrupt all jobs softly via ThreadInterrupted at safepoints
- if safepoint is not executed within some time period or it is but the job is still not cancelled, trigger a hard-interrupt by cancelling the job explicitly, if possible

Closes #11084.

### Important Notes

This change, while effectively does not alter the end result (ending jobs that take too much time), should reduce chances of IO Exceptions. It should also help identify where we lack safepoints polling or what java calls should be improved.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [ ] Unit tests have been written where possible.
- [ ] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

### CODE REVIEW

**Comment by hubertp** [comment]
Actually, the current solutions is wrong. Marking a Future as a (soft) cancel not only changes the state to `Cancelled` but also prevents further cancellations, potentially with interrupts, to be enacted.

File: engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/JobExecutionEngine.scala [hunk 1]
```diff
@@ -78,6 +84,84 @@ final class JobExecutionEngine(
...
+            val nextToCancel = pending
+              .filter { case (startTime, _) =>
+                startTime + forceInterruptTimeout > at
+              }
+              .sortBy(_._1)
+              .headOption
+            val timeout = nextToCancel
+              .map(j => Math.min(j._1 - at, forceInterruptTimeout))
+              .getOrElse(forceInterruptTimeout)
+            Thread.sleep(timeout)
```
Comment (JaroslavTulach) [inline engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/JobExecutionEngine.scala]: No `Thread.sleep` in production code, please.

File: engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/JobExecutionEngine.scala [hunk 2]
```diff
@@ -35,8 +35,14 @@ final class JobExecutionEngine(
   private val backgroundJobsRef =
     new AtomicReference[Vector[RunningJob]](Vector.empty)
 
+  private val pendingCancellations =
+    new AtomicReference[Vector[(Long, RunningJob)]](Vector.empty)
+
   private val context = interpreterContext.executionService.getContext
 
+  private val pendingCancellationsExecutor =
+    context.newFixedThreadPool(1, "pending-cancellations-timer", false)
```
Comment (JaroslavTulach) [inline engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/JobExecutionEngine.scala]: `EnsoContext.newFixedThreadPool` returns [ExecutorService](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/concurrent/ExecutorService.html). Can we wrap the default executor services with `SafepointAwareExecutorService` and whenever a method returning `Future` return a wrapper `SafepointAwareFuture` that would `cancel(true)` by asking `ThreadManager` first and only later emit an interrupt?

**Review by JaroslavTulach — COMMENTED** [review]
I have a feeling that by implementing our own `Future` + wrapping `ExecutionService` the code would be simpler.

Would it be possible to implement this without using new thread - e.g. `pendingCancellationsExecutor` at all? Possibly with own `Future.cancel`, right? It would just block until _safepoint_ or later `interrupt()` stops the `Future` computation.

I would feel more confident if the code was next to `EnsoContext`. First of all it would encapsulate all the logic and would be easier to unit test. Next, but not last: It would be written in more easily readable language.

**Comment by hubertp** [comment]
I got rid of `Thread.sleep` and removed the most controversial bits. I agree that would experiment with a more safepoint-aware execution in futures but I'd rather avoid implementing it in this PR. I created a separate ticket so that I don't forget our intentions for future enhancement.

**Review by 4e6 — APPROVED** [review]
> This change, while effectively does not alter the end result (ending jobs that take too much time), should reduce chances of IO Exceptions. It should also help identify where we lack safepoints polling or what java calls should be improved.

Ok, this makes sense

File: engine/runtime-integration-tests/src/test/scala/org/enso/interpreter/test/instrument/RuntimeAsyncCommandsTest.scala [hunk 3]
```diff
@@ -230,7 +230,7 @@ class RuntimeAsyncCommandsTest
     var iteration        = 0
     while (!isProgramStarted && iteration < 100) {
       val out = context.consumeOut
-      Thread.sleep(200)
+      Thread.sleep(100)
```
Comment (JaroslavTulach) [inline engine/runtime-integration-tests/src/test/scala/org/enso/interpreter/test/instrument/RuntimeAsyncCommandsTest.scala:233]: Common multithreaded pattern for awaiting output would be:
- modify `InterpreterContext._output` to have a subclass of `ByteArrayOutputStream` that notifies a monitor whenever something is written into it
- add `context.awaitOut(timeOut)` to check the output and `Object.wait(timeOut)` if the desired text hasn't yet arrived
- continue the test immediately when the desired text appears in the log by being notified as soon as possible

File: engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/job/Job.scala [hunk 4]
```diff
@@ -27,11 +29,32 @@ abstract class Job[+A](
     this(contextIds, isCancellable, mayInterruptIfRunning, false)
   }
 
+  /** Executes a job. Will mark the job as "started".
+    *
+    * @param ctx contains suppliers of services to perform a request
+    */
+  def run(implicit ctx: RuntimeContext): A = {
```
Comment (JaroslavTulach) [inline engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/job/Job.scala]: Should the behavior of `hasStated` be guaranteed, then this method has to be `final`.

File: engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/JobExecutionEngine.scala [hunk 5]
```diff
@@ -78,6 +80,82 @@ final class JobExecutionEngine(
...
+      extends Runnable {
+    private val forceInterruptTimeout: Long = 50 * 1000
+
+    override def run(): Unit = {
+      pendingJobs.sortBy(_._1).foreach {
+        case (timeRequestedToCancel, runningJob) =>
+          try {
+            val now                        = System.currentTimeMillis()
+            val timeSinceRequestedToCancel = now - timeRequestedToCancel
+            assert(timeSinceRequestedToCancel > 0)
```
Comment (JaroslavTulach) [inline engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/JobExecutionEngine.scala]: There is `assertInJVM` that is preferred to use when honoring `-ea` is the goal.

File: engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/JobExecutionEngine.scala [hunk 6]
```diff
@@ -78,6 +80,82 @@ final class JobExecutionEngine(
...
+              Level.FINEST,
+              "Job {} finished within the allocated soft-cancel time"
+            )
+          } catch {
+            case _: TimeoutException =>
+              val sb = new StringBuilder(
+                "Threaddump when timeout is reached while waiting for the job " + runningJob.id + " running in thread " + runningJob.job
+                  .threadNameExecutingJob() + " to cancel:\n"
+              )
+              Thread.getAllStackTraces.entrySet.forEach { entry =>
```
Comment (JaroslavTulach) [inline engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/JobExecutionEngine.scala:118]: Logging this kind of information may turn very useful.

**Review by JaroslavTulach — APPROVED** [review]

File: engine/runtime-integration-tests/src/test/scala/org/enso/interpreter/test/instrument/RuntimeAsyncCommandsTest.scala (hunk 3 above)
Comment (hubertp) [reply engine/runtime-integration-tests/src/test/scala/org/enso/interpreter/test/instrument/RuntimeAsyncCommandsTest.scala:233]: I left out the logic as it was defined before. We can probably improve it it becomes flaky.

**Pull request merged**